File: apps/backend/src/rhesis/backend/app/schemas/adaptive_testing.py
"""Schemas for adaptive testing API (generate outputs, evaluate, etc.)."""

from typing import Any, Dict, List, Optional

from pydantic import UUID4, BaseModel, Field

from rhesis.backend.app.schemas import Base
from rhesis.backend.app.schemas.test_set import TestSet as TestSetSchema

# ---------------------------------------------------------------------------
# Import from existing test set
# ---------------------------------------------------------------------------


class ImportAdaptiveTestSetResponse(Base):
    """Response for POST /adaptive_testing/import/{source_test_set_id}."""

    test_set: TestSetSchema
    imported: int = 0
    skipped: int = 0
    skipped_test_ids: List[str] = Field(default_factory=list)


class ExportAdaptiveTestSetResponse(Base):
    """Response for POST /adaptive_testing/export/{source_test_set_id}."""

    test_set: TestSetSchema
    exported: int = 0
    skipped: int = 0
    skipped_test_ids: List[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Generate outputs
# ---------------------------------------------------------------------------


class GenerateOutputsRequest(Base):
    """Request body for generating test outputs via endpoint invocation."""

    endpoint_id: Optional[UUID4] = None
    test_ids: Optional[List[UUID4]] = None
    topic: Optional[str] = None
    include_subtopics: bool = True
    overwrite: bool = False


class GenerateOutputsUpdatedItem(BaseModel):
    """One test whose output was updated (stored in test_metadata)."""

    test_id: str
    output: str


class GenerateOutputsFailedItem(BaseModel):
    """One test that failed during output generation."""

    test_id: str
    error: str


class GenerateOutputsResponse(Base):
    """Response for generate outputs. Results are stored in test metadata."""

    generated: int
    skipped: int = 0
    failed: List[GenerateOutputsFailedItem]
    updated: List[GenerateOutputsUpdatedItem]


# ---------------------------------------------------------------------------
# Evaluate
# ---------------------------------------------------------------------------


class EvaluateRequest(Base):
    """Request body for evaluating adaptive tests with specified metrics."""

    metric_names: Optional[List[str]] = None
    test_ids: Optional[List[UUID4]] = None
    topic: Optional[str] = None
    include_subtopics: bool = True
    overwrite: bool = False


class AdaptiveMetricEvalDetail(BaseModel):
    """Per-metric evaluation row (keyed by metric name on the parent model)."""

    score: float
    is_successful: bool
    reason: Optional[str] = None
    details: Optional[Dict[str, Any]] = None


class EvaluateResultItem(BaseModel):
    """One test that was evaluated successfully."""

    test_id: str
    label: str
    labeler: str
    model_score: float
    metrics: Optional[Dict[str, AdaptiveMetricEvalDetail]] = None


class EvaluateFailedItem(BaseModel):
    """One test that failed during evaluation."""

    test_id: str
    error: str


class EvaluateResponse(Base):
    """Response for evaluate. Results are persisted in test metadata."""

    evaluated: int
    skipped: int = 0
    results: List[EvaluateResultItem]
    failed: List[EvaluateFailedItem]


# ---------------------------------------------------------------------------
# Generate suggestions
# ---------------------------------------------------------------------------


class GenerateSuggestionsRequest(Base):
    """Request body for generating test suggestions via LLM."""

    topic: Optional[str] = None
    num_examples: int = Field(
        default=10,
        ge=1,
        le=50,
        description="Number of existing tests to sample as examples",
    )
    num_suggestions: int = Field(
        default=20,
        ge=1,
        le=100,
        description="Number of new test suggestions to generate",
    )
    user_feedback: Optional[str] = Field(
        default=None,
        max_length=1000,
        description="Optional user guidance to steer suggestion generation",
    )


class SuggestedTest(BaseModel):
    """One LLM-generated test suggestion (not yet persisted)."""

    topic: str = ""
    input: str = ""
    output: str = ""
    label: str = ""
    labeler: str = ""
    model_score: float = 0.0


class GenerateSuggestionsResponse(Base):
    """Response containing LLM-generated test suggestions."""

    suggestions: List[SuggestedTest]
    num_examples_used: int


# ---------------------------------------------------------------------------
# Generate suggestion outputs (non-persisted)
# ---------------------------------------------------------------------------


class SuggestionInput(BaseModel):
    """A single suggestion input for output generation."""

    input: str
    topic: str = ""


class GenerateSuggestionOutputsRequest(Base):
    """Request body for generating outputs for non-persisted suggestions."""

    endpoint_id: Optional[UUID4] = None
    suggestions: List[SuggestionInput]


class SuggestionOutputItem(BaseModel):
    """Result of generating output for one suggestion."""

    input: str
    output: str = ""
    error: Optional[str] = None


class GenerateSuggestionOutputsResponse(Base):
    """Response for suggestion output generation (nothing persisted)."""

    generated: int
    results: List[SuggestionOutputItem]


# ---------------------------------------------------------------------------
# Evaluate suggestions (non-persisted)
# ---------------------------------------------------------------------------


class SuggestionForEval(BaseModel):
    """A single suggestion to evaluate."""

    input: str
    output: str


class EvaluateSuggestionsRequest(Base):
    """Request body for evaluating non-persisted suggestions."""

    metric_names: Optional[List[str]] = None
    suggestions: List[SuggestionForEval]


class SuggestionEvalItem(BaseModel):
    """Evaluation result for one suggestion."""

    input: str
    label: str = ""
    labeler: str = ""
    model_score: float = 0.0
    metrics: Optional[Dict[str, AdaptiveMetricEvalDetail]] = None
    error: Optional[str] = None


class EvaluateSuggestionsResponse(Base):
    """Response for suggestion evaluation (nothing persisted)."""

    evaluated: int
    results: List[SuggestionEvalItem]


# ---------------------------------------------------------------------------
# Adaptive testing settings
# ---------------------------------------------------------------------------


class AdaptiveSettingsUpdate(Base):
    """Request body for updating adaptive testing settings.

    Both fields are optional so callers can update only what changed.
    """

    default_endpoint_id: Optional[UUID4] = None
    metric_ids: Optional[List[UUID4]] = None


class AdaptiveSettingsMetric(BaseModel):
    """Lightweight metric reference returned inside settings."""

    id: UUID4
    name: str


class AdaptiveSettingsEndpoint(BaseModel):
    """Lightweight endpoint reference returned inside settings."""

    id: UUID4
    name: str


class AdaptiveSettingsResponse(Base):
    """Response for GET/PUT adaptive testing settings."""

    default_endpoint: Optional[AdaptiveSettingsEndpoint] = None
    metrics: List[AdaptiveSettingsMetric] = []

File: apps/frontend/src/utils/api-client/interfaces/adaptive-testing.ts
/**
 * Interfaces for adaptive testing API.
 * Matches the backend schemas in apps/backend/src/rhesis/backend/app/schemas/adaptive_testing.py
 */

// =============================================================================
// Adaptive Test Set Interface
// =============================================================================

export interface AdaptiveTestSet {
  id: string;
  name: string;
  description?: string;
  slug?: string;
  nano_id?: string;
  status_id?: string;
  status?: string;
  test_set_type_id?: string;
  attributes?: Record<string, unknown>;
  created_at?: string;
  updated_at?: string;
}

/** Response from POST /adaptive_testing/import/{source_test_set_id} */
export interface ImportAdaptiveTestSetResponse {
  test_set: AdaptiveTestSet;
  imported: number;
  skipped: number;
  skipped_test_ids: string[];
}

/** Response from POST /adaptive_testing/export/{source_test_set_id} */
export interface ExportAdaptiveTestSetResponse {
  test_set: AdaptiveTestSet;
  exported: number;
  skipped: number;
  skipped_test_ids: string[];
}

// =============================================================================
// Test Node Interfaces
// =============================================================================

/** Per-metric evaluation row (tree/API key is metric name). */
export interface AdaptiveMetricEvalDetail {
  score: number;
  is_successful: boolean;
  reason?: string | null;
  details?: Record<string, unknown> | null;
}

export interface TestNode {
  id: string;
  topic: string;
  input: string;
  output: string;
  label: '' | 'topic_marker' | 'pass' | 'fail';
  labeler: string;
  to_eval: boolean;
  model_score: number;
  metrics?: Record<string, AdaptiveMetricEvalDetail> | null;
}

export interface TestNodeCreate {
  topic?: string;
  input?: string;
  output?: string;
  label?: '' | 'pass' | 'fail';
  labeler?: string;
  to_eval?: boolean;
  model_score?: number;
  /** When true, backend embeds test input and persists to embedding table */
  generate_embedding?: boolean;
}

export interface TestNodeUpdate {
  topic?: string;
  input?: string;
  output?: string;
  label?: '' | 'pass' | 'fail';
  to_eval?: boolean;
  model_score?: number;
}

// =============================================================================
// Topic Interfaces
// =============================================================================

export interface Topic {
  path: string;
  name: string;
  parent_path: string | null;
  depth: number;
  display_name: string;
  display_path: string;
  has_direct_tests: boolean;
  has_subtopics: boolean;
}

export interface TopicCreate {
  path: string;
  labeler?: string;
}

export interface TopicUpdate {
  new_name?: string;
  new_path?: string;
}

// =============================================================================
// Tree Interfaces
// =============================================================================

export interface TreeValidation {
  valid: boolean;
  missing_markers: string[];
  topics_with_tests: string[];
  topics_with_markers: string[];
}

export interface TreeStats {
  total_tests: number;
  total_topics: number;
  tests_by_topic: Record<string, number>;
}

// =============================================================================
// Response Interfaces
// =============================================================================

export interface DeleteTopicResponse {
  deleted: boolean;
  topic_path: string;
}

export interface DeleteTestResponse {
  deleted: boolean;
  test_id: string;
}

// =============================================================================
// Generate Outputs
// =============================================================================

export interface GenerateOutputsRequest {
  endpoint_id?: string | null;
  test_ids?: string[] | null;
  topic?: string | null;
  include_subtopics?: boolean;
  overwrite?: boolean;
}

export interface GenerateOutputsUpdatedItem {
  test_id: string;
  output: string;
}

export interface GenerateOutputsFailedItem {
  test_id: string;
  error: string;
}

export interface GenerateOutputsResponse {
  generated: number;
  skipped: number;
  failed: GenerateOutputsFailedItem[];
  updated: GenerateOutputsUpdatedItem[];
}

// =============================================================================
// Evaluate
// =============================================================================

export interface EvaluateRequest {
  metric_names?: string[] | null;
  test_ids?: string[] | null;
  topic?: string | null;
  include_subtopics?: boolean;
  overwrite?: boolean;
}

export interface EvaluateResultItem {
  test_id: string;
  label: string;
  labeler: string;
  model_score: number;
  metrics?: Record<string, AdaptiveMetricEvalDetail> | null;
}

export interface EvaluateFailedItem {
  test_id: string;
  error: string;
}

export interface EvaluateResponse {
  evaluated: number;
  skipped: number;
  results: EvaluateResultItem[];
  failed: EvaluateFailedItem[];
}

// =============================================================================
// Generate Suggestions
// =============================================================================

export interface GenerateSuggestionsRequest {
  topic?: string | null;
  num_examples?: number;
  num_suggestions?: number;
  user_feedback?: string | null;
  /** When true, API returns embedding vectors per suggestion input (not saved) */
  generate_embeddings?: boolean;
}

export interface SuggestedTest {
  topic: string;
  input: string;
  output: string;
  label: string;
  labeler: string;
  model_score: number;
  embedding?: number[] | null;
  /** Euclidean distance from centroid when generate_embeddings was true */
  diversity_score?: number | null;
}

export interface GenerateSuggestionsResponse {
  suggestions: SuggestedTest[];
  num_examples_used: number;
}

// =============================================================================
// Generate Suggestion Outputs (non-persisted)
// =============================================================================

export interface SuggestionInput {
  input: string;
  topic?: string;
}

export interface GenerateSuggestionOutputsRequest {
  endpoint_id?: string | null;
  suggestions: SuggestionInput[];
}

export interface SuggestionOutputItem {
  input: string;
  output: string;
  error?: string | null;
}

export interface GenerateSuggestionOutputsResponse {
  generated: number;
  results: SuggestionOutputItem[];
}

export interface SuggestionOutputStreamItemEvent {
  type: 'item';
  index: number;
  input: string;
  output: string;
  error: string | null;
}

export interface SuggestionOutputStreamSummaryEvent {
  type: 'summary';
  generated: number;
  total: number;
}

export type SuggestionOutputStreamEvent =
  | SuggestionOutputStreamItemEvent
  | SuggestionOutputStreamSummaryEvent;

// =============================================================================
// Evaluate Suggestions (non-persisted)
// =============================================================================

export interface SuggestionForEval {
  input: string;
  output: string;
}

export interface EvaluateSuggestionsRequest {
  metric_names?: string[] | null;
  suggestions: SuggestionForEval[];
}

export interface SuggestionEvalStreamItemEvent {
  type: 'item';
  index: number;
  input: string;
  label: string;
  labeler: string;
  model_score: number;
  metrics?: Record<string, AdaptiveMetricEvalDetail> | null;
  error: string | null;
}

export interface SuggestionEvalStreamSummaryEvent {
  type: 'summary';
  evaluated: number;
  total: number;
}

export type SuggestionEvalStreamEvent =
  | SuggestionEvalStreamItemEvent
  | SuggestionEvalStreamSummaryEvent;

// =============================================================================
// Adaptive Settings
// =============================================================================

export interface AdaptiveSettingsMetric {
  id: string;
  name: string;
}

export interface AdaptiveSettingsEndpoint {
  id: string;
  name: string;
}

export interface AdaptiveSettings {
  default_endpoint: AdaptiveSettingsEndpoint | null;
  metrics: AdaptiveSettingsMetric[];
}

export interface AdaptiveSettingsUpdateRequest {
  default_endpoint_id?: string | null;
  metric_ids?: string[] | null;
}

export interface SuggestionEvalItem {
  input: string;
  label: string;
  labeler: string;
  model_score: number;
  metrics?: Record<string, AdaptiveMetricEvalDetail> | null;
  error?: string | null;
}

export interface EvaluateSuggestionsResponse {
  evaluated: number;
  results: SuggestionEvalItem[];
}
